### PR TITLE
Add Recipe for using custom OIDC Idp

### PR DESCRIPTION
**Pull request by kwiatekus**
**Related issue(s)**
https://github.com/kyma-project/kyma/issues/11521

### CODE REVIEW

File: docs/03-tutorials/03-set-up-idp.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+kubectl get clusterrolebinding {BINDING_NAME}
+```
+
+To bind a group of users to the **kyma-admin** ClusterRole, run this command:
+
+```
+kubectl create clusterrolebinding {BINDING_NAME} --clusterrole=kyma-admin --group={GROUPNAME}
+```
+
+You can combine user and group level permission in one binding. Run `kubectl create clusterrolebinding --help` in your terminal for more options.
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:84]:
```suggestion
You can combine user and group level permissions in one binding. Run `kubectl create clusterrolebinding --help` in your terminal for more options.
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+### Configure kubectl access
+
+With this step you will set up the OIDC provider in  kubeconfig file to enforce authentication flow when accessing kyma via `kubectl`.
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:88]:
```suggestion
With this step, you will set up the OIDC provider in the kubeconfig file to enforce authentication flow when accessing Kyma using `kubectl`.
```

Comment (kazydek) [reply docs/03-tutorials/03-set-up-idp.md:88]: Rather "using" than "via -> https://kyma-project.io/community/guidelines/content/#style-and-terminology-style-and-terminology-terminology

File: docs/03-tutorials/03-set-up-idp.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+1. Install [kubelogin](https://github.com/int128/kubelogin) plugin
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:90]:
```suggestion
1. Install the [kubelogin](https://github.com/int128/kubelogin) plugin.
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+2. Copy your current kubeconfig file into a new file
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:91]:
```suggestion
2. Copy your current kubeconfig file into a new file.
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+3. In the new kubeconfig file define a new `oidc` user and setup OIDC provider, as follows:
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:92]:
```suggestion
3. In the new kubeconfig file, define a new OIDC user and set up the OIDC provider. 
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        apiVersion: client.authentication.k8s.io/v1beta1
+        command: kubectl
+        args:
+        - oidc-login
+        - get-token
+        - --oidc-issuer-url=ISSUER_URL
+        - --oidc-client-id=YOUR_CLIENT_ID
+        #- --oidc-client-secret=YOUR_CLIENT_SECRET this is not required if your OICS server supports PKCE authentication flow
+    ```
+4. To enforce oidc login set the `oidc` user as a default user in the context
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:108]:
```suggestion
4. To enforce the OIDC login, set the OIDC user as a default user in the context.
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    user:
+        exec:
+        apiVersion: client.authentication.k8s.io/v1beta1
+        command: kubectl
+        args:
+        - oidc-login
+        - get-token
+        - --oidc-issuer-url=ISSUER_URL
+        - --oidc-client-id=YOUR_CLIENT_ID
+        #- --oidc-client-secret=YOUR_CLIENT_SECRET this is not required if your OICS server supports PKCE authentication flow
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:106]:
```suggestion
        #- --oidc-client-secret=YOUR_CLIENT_SECRET this is not required if your OICS server supports the PKCE authentication flow
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    ```
+4. To enforce oidc login set the `oidc` user as a default user in the context
+    ```yaml
+    context:
+        cluster: {your cluster name}
+        user: oidc
+    ```
+5. Now you can share the modified kubeconfig file to members of your team / organisation. When they will use it, your identity provider will do the authentication for you and the kubernetes api server will make sure they will have access to resources according to the role bound to them as individuals or group members.
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:114]:
```suggestion
5. Now you can share the modified kubeconfig file with the members of your team or organization. When they use it, your identity provider will handle the authentication. The Kubernetes API server will make sure they will have access to resources according to the role bound to them as individuals or group members.     
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        - oidc-login
+        - get-token
+        - --oidc-issuer-url=ISSUER_URL
+        - --oidc-client-id=YOUR_CLIENT_ID
+        #- --oidc-client-secret=YOUR_CLIENT_SECRET this is not required if your OICS server supports PKCE authentication flow
+    ```
+4. To enforce oidc login set the `oidc` user as a default user in the context
+    ```yaml
+    context:
+        cluster: {your cluster name}
```
Comment (kazydek) [inline docs/03-tutorials/03-set-up-idp.md:111]:
```suggestion
        cluster: {YOUR_CLUSTER_NAME}
```

File: docs/03-tutorials/03-set-up-idp.md [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+```
+
+If you use managed kubernetes you will do it different depending on your provider.
+For example, if you use gardener as a managed kubernetes offering you will probably wnt to look at the [OIDCPreset](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md) resources that help you with the task.
+
+
+
+### Configure Role Based Access to identies provided by your OIDC server
+
+Now, whenever api server is called with a JWT token, the api server will be able to validate and extract the associated identity ( from the `username` and `group` claims of the JWT token).
```
[the remark this answers is not in the document]
Comment (kwiatekus) [reply docs/03-tutorials/03-set-up-idp.md:55]: Not really JWT token calling api server. I will rephrase.

File: docs/03-tutorials/03-set-up-idp.md [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+--k3s-server-arg "--kube-apiserver-arg=oidc-issuer-url=<your-ipd-issuer-url>" \
+--k3s-server-arg "--kube-apiserver-arg=oidc-username-claim=<username-claim-at-your-ipd>" \
+--k3s-server-arg "--kube-apiserver-arg=oidc-client-id=<your-ipd-client-id>" \
+--k3s-server-arg "--kube-apiserver-arg=oidc-groups-claim=<group-claim-at-your-ipd>" \
+```
+
+If you use managed kubernetes you will do it different depending on your provider.
+For example, if you use gardener as a managed kubernetes offering you will probably wnt to look at the [OIDCPreset](https://github.com/gardener/gardener/blob/master/docs/usage/openidconnect-presets.md) resources that help you with the task.
+
+
```
[the remark this answers is not in the document]
Comment (kwiatekus) [reply docs/03-tutorials/03-set-up-idp.md]: done

File: docs/03-tutorials/03-set-up-idp.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (kwiatekus) [reply docs/03-tutorials/03-set-up-idp.md]: done

File: docs/03-tutorials/03-set-up-idp.md [hunk 12]
```diff
@@ -0,0 +1,114 @@
+---
+title: Set up a custom identity provider
+type: Tutorials
```
Comment (NHingerl) [inline docs/03-tutorials/03-set-up-idp.md:3]:
```suggestion
```

`type` metadata is not used any longer, right?

File: docs/03-tutorials/03-set-up-idp.md [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+- `clientSecret` 
+
+2. Add `http://localhost:8000` to allowed redirect URIs that are required for the OIDC login plugin.
+3. Configure the name of the `username` and `group` claims.
+4. Enable the Proof Key for Code Exchange (PKCE) authentication flow.
+
+### Configure your identity provider as the OIDC server
+
+In general, you have to add flags to the API server as described in the [Kubernetes documentation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#configuring-the-api-server). You will do it in different ways depending on your Kubernetes distribution.
+For example, if you want to use  k3d, you need to pass additional `--k3s-server-arg` flags containing the OIDC server configuration when creating the cluster. See the [specification](https://k3d.io/usage/commands/k3d_cluster_create/) of the `k3d cluster create` command:
```
Comment (NHingerl) [inline docs/03-tutorials/03-set-up-idp.md:38]:
```suggestion
For example, if you want to use k3d, you need to pass additional `--k3s-server-arg` flags containing the OIDC server configuration when creating the cluster. See the [specification](https://k3d.io/usage/commands/k3d_cluster_create/) of the `k3d cluster create` command:
```

**Review by NHingerl — CHANGES_REQUESTED** [review]

**Review by NHingerl — APPROVED** [review]

**Pull request merged**